streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -623,63 +642,75 @@ void runOnce() {
+             *  1. If it is time to punctuate, do it;
+             *  2. If it is time to commit, do it, this should be after 1) since punctuate may trigger commit;
+             *  3. If there's no records processed, end the current iteration immediately;
+             *  4. If we are close to consumer's next poll deadline, end the current iteration immediately;
+             *  5. If any of 1), 2) and 4) happens, half N for next iteration;
+             *  6. Otherwise, increment N.
              */
-            int processed = 0;
-            long timeSinceLastPoll;
-
             do {
-                for (int i = 0; i < numIterations; i++) {
-                    advanceNowAndComputeLatency();
-                    processed = taskManager.process(now);
-
-                    if (processed > 0) {
-                        // It makes no difference to the outcome of these metrics when we record "0",
-                        // so we can just avoid the method call when we didn't process anything.
-                        processRateSensor.record(processed, now);
-
-                        // This metric is scaled to represent the _average_ processing time of _each_
-                        // task. Note, it's hard to interpret this as defined, but we would need a KIP
-                        // to change it to simply report the overall time spent processing all tasks.
-                        final long processLatency = advanceNowAndComputeLatency();
-                        processLatencySensor.record(processLatency / (double) processed, now);
-
-                        // commit any tasks that have requested a commit
-                        final int committed = taskManager.maybeCommitActiveTasksPerUserRequested();
-
-                        if (committed > 0) {
-                            final long commitLatency = advanceNowAndComputeLatency();
-                            commitSensor.record(commitLatency / (double) committed, now);
-                        }
-                    } else {
-                        // if there is no records to be processed, exit immediately
-                        break;
-                    }
+                final int processed = taskManager.process(numIterations, now);
+                if (processed > 0) {
+                    // It makes no difference to the outcome of these metrics when we record "0",
+                    // so we can just avoid the method call when we didn't process anything.
+                    processRateSensor.record(processed, now);
+
+                    // This metric is scaled to represent the _average_ processing time of _each_
+                    // task. Note, it's hard to interpret this as defined, but we would need a KIP
+                    // to change it to simply report the overall time spent processing all tasks.
+                    final long processLatency = advanceNowAndComputeLatency();
```

Comment:
I noticed that when preparing the PR, and I was biting this with the hope to reduce calls to `time.milliseonds` inside `advanceNowAndComputeLatency` when `processed / punctuated / committed == 0`, with the assumption that the overhead is small and hence no need to worry.

But after thinking it twice as you brought it up as well, I think it might be better to not be stingy about `time.milliseconds` calls but rather trying to be more accurate in our latency recordings. So I will go ahead and change this.